src/tests/test_runtime_env.py

```diff
@@ -0,0 +1,550 @@
+                assert not RUNTIME_ENV.IN_BROWSER, "Should not detect browser in Shell"
+                assert not RUNTIME_ENV.IN_NODE, "Should not detect Node.js in Shell"
+                assert not RUNTIME_ENV.IN_DENO, "Should not detect Deno in Shell"
+                assert not RUNTIME_ENV.IN_BUN, "Should not detect Bun in Shell"
+
+                # Shell-specific flags should be false
+                assert not RUNTIME_ENV.IN_NODE_COMMONJS, "Shell should not be CommonJS"
+                assert not RUNTIME_ENV.IN_NODE_ESM, "Shell should not be Node ESM"
+                assert not RUNTIME_ENV.IN_BROWSER_MAIN_THREAD, "Shell should not be browser main thread"
+                assert not RUNTIME_ENV.IN_BROWSER_WEB_WORKER, "Shell should not be web worker"
+            else:
+                # If not in Shell, just verify the detection is consistent
+                assert RUNTIME_ENV.IN_BROWSER or RUNTIME_ENV.IN_NODE or RUNTIME_ENV.IN_DENO or RUNTIME_ENV.IN_BUN, \
+                    "Should be in some runtime environment"
+        """)
+
+
+class TestBackwardCompatibility:
+    """Backward compatibility tests."""
+
+    def test_runtime_env_singleton_behavior(self, selenium, runtime_flags):
+        """Test RUNTIME_ENV singleton behavior."""
+        selenium.run(f"""
+            from pyodide_js import RUNTIME_ENV as env1
+            from pyodide_js import RUNTIME_ENV as env2
+
+            required_flags = {runtime_flags!r}
+
+            for flag in required_flags:
+                value1 = getattr(env1, flag)
+                value2 = getattr(env2, flag)
+                assert isinstance(value1, bool), f"{{flag}} should be boolean, got {{type(value1)}}"
+                assert isinstance(value2, bool), f"{{flag}} should be boolean, got {{type(value2)}}"
+                assert value1 == value2, (
+                    f"RUNTIME_ENV.{{flag}} should be consistent across imports "
+                    f"(got {{value1}} vs {{value2}})"
+                )
+        """)
+
+    def test_runtime_env_accessible_from_python(self, selenium):
+        """Test RUNTIME_ENV Python accessibility."""
+        selenium.run("""
+            from pyodide_js import RUNTIME_ENV
+
+            # Should be able to access flags directly
+            browser_flag = RUNTIME_ENV.IN_BROWSER
+            node_flag = RUNTIME_ENV.IN_NODE
+
+            # Should be boolean values
+            assert isinstance(browser_flag, bool)
+            assert isinstance(node_flag, bool)
+
+            # Should be mutually exclusive in most cases
+            if browser_flag:
+                assert not node_flag, "Cannot be both browser and Node.js"
+            if node_flag:
+                assert not browser_flag, "Cannot be both Node.js and browser"
+        """)
+
+
+class TestWebWorkerEnvironment:
+    """Web worker environment tests."""
+
+    @pytest.mark.xfail_browsers(
+        chrome="webworker only", firefox="webworker only", safari="webworker only"
+    )
```

Comment:
```suggestion
    @pytest.mark.xfail_browsers(
        safari="safari web workers are flaky"
    )
```

Webworker test should test chrome and firefox.